hail/src/main/scala/is/hail/expr/ir/streams/EmitStream.scala

```diff
@@ -257,6 +261,147 @@ object EmitStream {
+                dictState.withContainer(cb, resultKeyValue, { cb =>
+                  emitVoid(seqOps, cb, container = Some(keyedContainer), env = env.bind(name, eltField))
+                })
+                cb.ifx(dictState.size >= maxSize,{
+                  cb.assign(produceElementMode, true)
+                })
+
+                cb.ifx(produceElementMode, {
+                  cb.goto(elementProduceLabel)},
+                  {
+                  cb.goto(getElemLabel)
+                  }
+                )
+
+                cb.define(childProducer.LendOfStream)
+                cb.assign(childStreamEnded, true)
+                cb.assign(produceElementMode, true)
+
+                cb.define(elementProduceLabel)
+                cb.ifx(numElemInArray ceq 0, {
+                  dictState.tree.foreach(cb) { (cb, elementOff) =>
+                    cb += nodeArray.update(numElemInArray, elementOff)
+                    cb.assign(numElemInArray, numElemInArray + 1)
+                  }
+                })
+
+                cb.ifx(numElemInArray <= idx, {
+                  cb.assign(idx, 0)
+                  cb.assign(numElemInArray, 0)
+                  cb.assign(produceElementMode, false)
+                  cb.ifx(childStreamEnded , {
+                    cb.goto(LendOfStream)
+                  }, {
+                    cb.goto(startLabel)
+                  })
+                })
+                val nodeAddress = cb.memoize(nodeArray(idx))
+                cb.assign(idx, idx + 1)
+                dictState.loadNode(cb, nodeAddress)
+                val serializedAggValue = keyedContainer.container.states.states.map(state => state.serializeToRegion(cb, PCanonicalBinary(), region))
+                val key = dictState.keyed.storageType.loadCheapSCode(cb, nodeAddress).loadField(cb, "kt").memoize(cb, "steam_buff_agg_key")
+                val serializedAggEmitCodes = serializedAggValue.map(aggValue => EmitCode.present(mb, aggValue))
+                val serializedAggTupleSValue = SStackStruct.constructFromArgs(cb, region, serializedAggSType.virtualType, serializedAggEmitCodes: _*)
+                val keyValue = key.get(cb).asInstanceOf[SBaseStructValue]
+                val sStructToReturn = keyValue.insert(cb, region, returnElemType.asInstanceOf[TStruct], ("agg", EmitCode.present(mb, serializedAggTupleSValue)
+                  .memoize(cb, "stream_buff_agg_return_val")))
+                assert(returnElemSType.virtualType == sStructToReturn.st.virtualType)
+                val casted = sStructToReturn.castTo(cb, region, returnElemSType)
+                cb.assign(newStreamElem, EmitCode.present(mb, casted).toI(cb))
+                cb.goto(LproduceElementDone)
+              }
+
+              override val element: EmitCode = newStreamElem.load
+
+              override def close(cb: EmitCodeBuilder): Unit = {
+                childProducer.close(cb)
```

Comment:
add:
```
                if (childProducer.requiresMemoryManagementPerElement)
                  cb += childProducer.elementRegion.invalidate()
```

Also, needs `cb += dictState.region.invalidate()` (the perennial foot-nuke)